File: Baekjoon/[ 15961 ] 회전 초밥.py
# https://www.acmicpc.net/problem/15961
# 접시수, 초밥의 가짓수, 연속 접시수, 쿠폰번호
import sys
I = sys.stdin.readline
N, D, K, C = list(map(int, I().split()))
arr = [int(I()) for _ in range(N)]
counts = [0] * (D+1)
count = 0

for i in range(K):
    counts[arr[i]] += 1

for i in range(1,D+1):
    if counts[i] > 0:
        count += 1

if counts[C] == 0:
    answer = count+1
else:
    answer = count

for i in range(K, N+K):
    counts[arr[i%N]] += 1
    if counts[arr[i%N]] == 1:
        count += 1
    
    counts[arr[i-K]] -= 1
    if counts[arr[i-K]] == 0:
        count -= 1
    
    if counts[C] == 0:
        answer = max(answer, count+1)
    else:
        answer = max(answer, count)

print(answer)

File: Baekjoon/[ 15961 ] 회전 초밥.java
// https://www.acmicpc.net/problem/15961
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

	static int N, d, k, c, answer;
	static int[] chobab;
	static int[] D;
	
	public static void main(String[] args) throws IOException {
		// 1. 입력
		BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(in.readLine(), " ");
		N = Integer.parseInt(st.nextToken()); // 접시 수		 	3,000,000
		d = Integer.parseInt(st.nextToken()); // 초밥 종류 수		3000
		k = Integer.parseInt(st.nextToken()); // 연속해서 먹는 접시 수	3000
		c = Integer.parseInt(st.nextToken()); // 쿠폰 번호			3000
		chobab = new int[N]; // 초밥 배열
		for(int i=0; i<N; i++) {
			chobab[i] = Integer.parseInt(in.readLine());
		}
		
		// 2. 풀이
		D = new int[d+1]; // 초밥 종류
		for(int i=0; i<k; i++) {
			D[chobab[i]]++; // 0 0 0 0 0 0 0 2 0 1 0 0 0 0 0 0 0 1 0
		}
		D[c] ++;
		int count = 0;
		for(int i=0; i<d+1; i++) {
			if(D[i] != 0) count++;  // 먹은 종류 갯수
		}
		
		int max = count;
		for(int i=0; i<N; i++) {
			D[chobab[i]]--;
			if(D[chobab[i]] == 0) count--;
			
			if(D[chobab[(i+k)%N]] == 0) count++;
			D[chobab[(i+k)%N]]++;
			
			max = Math.max(max, count);
		}
		
		// 3. 최대값 출력
		System.out.println(max);
	}
}